### ISSUE TITLE
Brown noise looks different with different sample length

### ISSUE THREAD

**Issue by tomato18463** [felixpatzelt/colorednoise#16]
Hi, 

If I generate a brownian noise with 16000 sample points using your library (something like `y = powerlaw_psd_gaussian(2.0, 16000)`), and save it with a sample rate of 16000, its fft spectrum looks as follows:

![no title](https://github.com/felixpatzelt/colorednoise/assets/113895603/93c24cf1-c692-4f4c-8431-089bd0486ff3)

If I generate the same noise with 16000000 samples and save it using the same sample rate as before (16000), the fft spectrum looks like:

![no title1](https://github.com/felixpatzelt/colorednoise/assets/113895603/4b86a9c2-382a-45e6-97aa-c05f0d78d2ec)

As you can see, the spectrum looks quite different, and they actually sound very different. Is there anything wrong with my way of using your code? Also, as you can see in the second figure, there is a drastic drop in the wave plot near the end of the wave. It seems the wave drop to a very low value and comes back immediately in the point after. Is this normal?

Thanks for the help!

**Comment by felixpatzelt** [comment]
This is indeed expected behaviour because the library. 

To understand why, you first need to realise, that some long-range correlated signal may have a diverging variance or even mean. That is, the longer the signal becomes, the bigger the value you measure will be. The variance starts to diverge when the power spectrum decays like 1 / f or faster. For brownian noise the power spectrum decays like 1 / f^2, so there the variance also diverges. See also here: https://en.wikipedia.org/wiki/Power_law

Intuitively for a brownian noise (which could also be called a brownian motion) can be understood as your signal wandering arbitrarily far away from where it started if your generate enough time steps. As soon as you start limiting the signal to stay in a certain range for sure, you no longer have a pure brownian noise.

As you can see, this creates a problem for many practical applications. Normally we want a signals range to be controlled. That's why in this library, the signal is always rescaled to have (on average over many runs) a variance of one - no matter what the parameters are you give the signal. It is then the users choice to scale the signal to the desired level.

There are two possible ways in this library to not have the output variance change with the signal length:
- Choose an exponent between zero and one
- Chose an a low-frequency cutoff fmin > 0.

The latter creates a broken power law that is only valid for frequencies above the cutoff. Below the cutoff, the signal will have a flat power spectrum like white noise. This limits large excursions of the signal over very long times. See docstring of powerlaw_psd_gaussian for more explanations on this.

**Comment by felixpatzelt** [comment]
About the second part of your question: The "drastic drop" in your second time series plot looks indeed strange. This should normally not happen. Are you sure output of the library already included this artefact or might it have been caused by later processing or the plotting function?

**Comment by tomato18463** [comment]
> About the second part of your question: The "drastic drop" in your second time series plot looks indeed strange. This should normally not happen. Are you sure output of the library already included this artefact or might it have been caused by later processing or the plotting function?

This is created by the following code:

```
import numpy as np
from scipy.io import wavfile
y = powerlaw_psd_gaussian(2.0, 16000000)
y_ = y / np.abs(y).max()
wavfile.write("tmp.wav", 16000, (32768*y_).astype(np.int16))
```

where `powerlaw_psd_gaussian` is the function defined [here](https://github.com/felixpatzelt/colorednoise/blob/master/colorednoise.py). It can be also seen if you plot the generated signal using plt, so I don't think its the problem of plotting function.

**Comment by felixpatzelt** [comment]
A signed 16 bit integer has a range of `-32768` to `+32767`. So if you multiply by `32767` instead of `32768` in the last line of the code you posted the problem should be gone. You can check the range using e.g. `np.iinfo(np.int16)`.

**Comment by tomato18463** [comment]
> This change in the spectrum is indeed expected behaviour.
> 
> To understand why, you first need to realise that some long-range correlated signals may not be stationary. That is, the longer the signal becomes, the bigger the variance will be. This starts when the power spectrum decays like 1 / f or faster. For brownian noise the power spectrum decays like 1 / f^2, so this signal is also not stationary.
> 
> Intuitively for a brownian noise (which may also be called a 1-dimensional brownian motion or random walk, or Wiener process) this can be understood as your signal wandering arbitrarily far away from where it started if your generate enough steps. As soon as you would start hard-limiting the signal to stay in a certain range for sure, you no longer have a pure brownian noise.
> 
> As you can see, this creates a problem for many practical applications. Normally we want a signal's range to be controlled. That's why in this library, the signal is always rescaled to have (on average over many runs) a variance of one - no matter what parameters you give to the function generating the signal. In simpler terms, the output is scaled down according to the number of samples that are generated. It is then the user's choice to scale the signal to the desired level afterwards.
> 
> There are two possible ways in this library to not have the output be rescaled differently for different-length signals:
> 
> * Choose an exponent between zero and one
> * Chose an a low-frequency cutoff fmin > 0.
> 
> The latter creates a "broken" power law that only follows the exponent for frequencies above the cutoff. Below the cutoff, the signal will have a flat power spectrum like white noise. This limits large excursions of the signal over very long times. E.g. for an audio signal you could set the cutoff to be below the hearing range. This is what a first order lowpass filter or an Ornstein-Uhlenbeck-Process (also called leaky integrator) would produce. See docstring of powerlaw_psd_gaussian for more explanations on the fmin parameter.
> 
> See also https://en.wikipedia.org/wiki/Brownian_noise https://en.wikipedia.org/wiki/Wiener_process#Properties_of_a_one-dimensional_Wiener_process https://en.wikipedia.org/wiki/Long-range_dependence

I see. Thanks for your detailed explanation!

**Comment by tomato18463** [comment]
> A signed 16 bit integer has a range of `-32768` to `+32767`. So if you multiply by `32767` instead of `32768` in the last line of the code you posted the problem should be gone. You can check the range using e.g. `np.iinfo(np.int16)`.

Yes the problem is gone. Thank you!
